### PR TITLE
CPU Performance Improvements

### PR DESCRIPTION
**Pull request by smitdylan2001**
No changes were made a change to functionality, only performance improvement.

- Use simpler return statements in WebXRManager & WebXRPackage (very small improvement)
- Use .CompareTag instead of string comparison in ControllerInteraction (no garbage allocation)
- Use TryGetComponent in Controller Interaction (no garbage allocation when component is null)
- Use SetLocalPositionAndRotation instead of 2 separate calls. This changes the amount of time a hierarchy needs to be calculated from 2 to 1, improving CPU substantially for big hierarchies. 

transform.SetLocalPositionAndRotation is the biggest changeAdded in [2021.3.11](https://unity.com/releases/editor/whats-new/2021.3.11), but since there is no compilation check except #if UNITY_2021_3_OR_NEWER, people using version .0 to .10 would get compilation errors. 
#if UNITY_2021_3_OR_NEWER can be used since updates within the same LTS version rarely break the project, but for safety not included in the PR yet.

### CODE REVIEW

**Comment by De-Panther** [comment]
Awesome! Thanks!

File: Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs [hunk 1]
```diff
@@ -183,19 +183,17 @@ public void SetUseCollidersForHandJoints(bool value)
...
         }
 #if WEBXR_INPUT_PROFILES
         if (handModelJoints.ContainsKey(i))
         {
-          var collider = handModelJoints[i].GetComponent<Collider>();
-          if (collider != null)
+          if (handModelJoints[i].TryGetComponent<Collider>(out var collider))
           {
-            collider.enabled = useCollidersForHandJoints;
+                        collider.enabled = useCollidersForHandJoints;
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs]:
```suggestion
              collider.enabled = useCollidersForHandJoints;
```

File: Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs [hunk 2]
```diff
@@ -259,11 +257,16 @@ private void SetControllerVisible(bool visible)
 
     private void SetInputProfileModelPose(bool alwaysUseGrip)
     {
+#if UNITY_2022_3_OR_NEWER
+      inputProfileModelParent.transform.SetLocalPositionAndRotation(alwaysUseGrip ? Vector3.zero : controller.gripPosition, 
+          alwaysUseGrip ? Quaternion.identity : controller.gripRotation);
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs]:
```suggestion
        alwaysUseGrip ? Quaternion.identity : controller.gripRotation);
```

File: Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs [hunk 3]
```diff
@@ -259,11 +257,16 @@ private void SetControllerVisible(bool visible)
...
+#else
       inputProfileModelParent.transform.localPosition = alwaysUseGrip ? Vector3.zero : controller.gripPosition;
       inputProfileModelParent.transform.localRotation = alwaysUseGrip ? Quaternion.identity : controller.gripRotation;
-    }
+#endif
+        }
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs]:
```suggestion
    }
```

File: Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs [hunk 4]
```diff
@@ -259,11 +257,16 @@ private void SetControllerVisible(bool visible)
...
 
-    private void SetHandJointsVisible(bool visible)
+        private void SetHandJointsVisible(bool visible)
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs]:
```suggestion
    private void SetHandJointsVisible(bool visible)
```

File: Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs [hunk 5]
```diff
@@ -328,18 +331,26 @@ private void OnHandUpdate(WebXRHandData handData)
...
         if (handJoints.ContainsKey(i))
         {
+#if UNITY_2022_3_OR_NEWER
+          handJoints[i].SetLocalPositionAndRotation(rotationOffset * (handData.joints[i].position - handData.joints[0].position), rotationOffset * handData.joints[i].rotation);
+#else
           handJoints[i].localPosition = rotationOffset * (handData.joints[i].position - handData.joints[0].position);
           handJoints[i].localRotation = rotationOffset * handData.joints[i].rotation;
-          if (handData.joints[i].radius != handJoints[i].localScale.x && handData.joints[i].radius > 0)
+#endif
+        if (handData.joints[i].radius != handJoints[i].localScale.x && handData.joints[i].radius > 0)
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs]:
```suggestion
          if (handData.joints[i].radius != handJoints[i].localScale.x && handData.joints[i].radius > 0)
```

File: Packages/webxr-interactions/Runtime/Scripts/SceneHitTest.cs [hunk 6]
```diff
@@ -77,8 +77,12 @@ void Update()
 
     private void HandleOnXRChange(WebXRState state, int viewsCount, Rect leftRect, Rect rightRect)
     {
-      WebXRManager.Instance.transform.localPosition = originPosition;
+#if UNITY_2022_3_OR_NEWER
+       WebXRManager.Instance.transform.SetLocalPositionAndRotation(originPosition, originRotation);
+#else
+       WebXRManager.Instance.transform.localPosition = originPosition;
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/SceneHitTest.cs]:
```suggestion
#if UNITY_2022_3_OR_NEWER
      WebXRManager.Instance.transform.SetLocalPositionAndRotation(originPosition, originRotation);
#else
      WebXRManager.Instance.transform.localPosition = originPosition;
```

File: Packages/webxr-interactions/Runtime/Scripts/SceneHitTest.cs [hunk 7]
```diff
@@ -97,26 +101,39 @@ void HandleOnViewerHitTestUpdate(WebXRHitPoseData hitPoseData)
...
       float angle = Mathf.Atan2(diff.y, diff.x) * Mathf.Rad2Deg - 90f;
       Quaternion rotationOffset = Quaternion.Euler(0, angle, 0);
+#if UNITY_2022_3_OR_NEWER
+      WebXRManager.Instance.transform.SetLocalPositionAndRotation(rotationOffset * (originPosition - hitPoseData.position), rotationOffset);
+#else
       WebXRManager.Instance.transform.localPosition = rotationOffset * (originPosition - hitPoseData.position);
       WebXRManager.Instance.transform.localRotation = rotationOffset;
+#endif
+
+    }
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/SceneHitTest.cs]:
```suggestion
```

File: Packages/webxr-interactions/Runtime/Scripts/SceneHitTest.cs [hunk 8]
```diff
@@ -97,26 +101,39 @@ void HandleOnViewerHitTestUpdate(WebXRHitPoseData hitPoseData)
...
     }
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/SceneHitTest.cs:136]:
```suggestion
    }
  }
```

File: Packages/webxr/Runtime/Scripts/WebXRCamera.cs [hunk 9]
```diff
@@ -89,21 +89,31 @@ private void UpdateFollower()
       switch (xrState)
       {
         case WebXRState.AR:
+#if UNITY_2022_3_OR_NEWER
+          cameraFollower.SetLocalPositionAndRotation(viewsCount > 1 ? (cameraARL.transform.localPosition + cameraARR.transform.localPosition) * 0.5f : cameraARL.transform.localPosition,
+               cameraARL.transform.localRotation);
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRCamera.cs]:
```suggestion
            cameraARL.transform.localRotation);
```

File: Packages/webxr/Runtime/Scripts/WebXRCamera.cs [hunk 10]
```diff
@@ -89,21 +89,31 @@ private void UpdateFollower()
...
-            return;
-          }
-          cameraFollower.localPosition = cameraARL.transform.localPosition;
+#endif
           return;
         case WebXRState.VR:
-          cameraFollower.localRotation = cameraL.transform.localRotation;
+#if UNITY_2022_3_OR_NEWER
+          cameraFollower.SetLocalPositionAndRotation((cameraL.transform.localPosition + cameraR.transform.localPosition) * 0.5f, 
+              cameraL.transform.localRotation);
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRCamera.cs]:
```suggestion
            cameraL.transform.localRotation);
```

File: Packages/webxr/Runtime/Scripts/WebXRCamera.cs [hunk 11]
```diff
@@ -89,21 +89,31 @@ private void UpdateFollower()
...
       }
-      cameraFollower.localRotation = cameraMain.transform.localRotation;
-      cameraFollower.localPosition = cameraMain.transform.localPosition;
+#if UNITY_2022_3_OR_NEWER
+        cameraFollower.SetLocalPositionAndRotation(cameraMain.transform.localPosition, 
+            cameraMain.transform.localRotation);
+#else
+        cameraFollower.localRotation = cameraMain.transform.localRotation;
+        cameraFollower.localPosition = cameraMain.transform.localPosition;
+#endif
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRCamera.cs]:
```suggestion
#if UNITY_2022_3_OR_NEWER
      cameraFollower.SetLocalPositionAndRotation(cameraMain.transform.localPosition, 
        cameraMain.transform.localRotation);
#else
      cameraFollower.localRotation = cameraMain.transform.localRotation;
      cameraFollower.localPosition = cameraMain.transform.localPosition;
#endif
```

File: Packages/webxr/Runtime/Scripts/WebXRCamera.cs [hunk 12]
```diff
@@ -168,21 +178,38 @@ private void OnXRChange(WebXRState state, int viewsCount, Rect leftRect, Rect ri
         Vector3 rightPosition)
     {
       if (xrState == WebXRState.VR)
-      {
+            {
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRCamera.cs]:
```suggestion
      {
```

File: Packages/webxr/Runtime/Scripts/WebXRCamera.cs [hunk 13]
```diff
@@ -168,21 +178,38 @@ private void OnXRChange(WebXRState state, int viewsCount, Rect leftRect, Rect ri
...
+        cameraR.transform.SetLocalPositionAndRotation(rightPosition, rightRotation);
+#else
         cameraR.transform.localPosition = rightPosition;
         cameraR.transform.localRotation = rightRotation;
+#endif
         cameraR.projectionMatrix = rightProjectionMatrix;
       }
       else if (xrState == WebXRState.AR)
-      {
+            {
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRCamera.cs]:
```suggestion
      {
```

File: Packages/webxr/Runtime/Scripts/WebXRController.cs [hunk 14]
```diff
@@ -371,13 +371,23 @@ private void OnControllerUpdate(WebXRControllerData controllerData)
           gripPosition = controllerData.gripPosition;
           if (alwaysUseGrip)
           {
+#if UNITY_2022_3_OR_NEWER
+            transform.SetLocalPositionAndRotation(controllerData.rotation * (controllerData.position + controllerData.gripPosition), 
+                controllerData.rotation * controllerData.gripRotation);
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRController.cs]:
```suggestion
              controllerData.rotation * controllerData.gripRotation);
```

File: Packages/webxr/Runtime/Scripts/WebXRController.cs [hunk 15]
```diff
@@ -371,13 +371,23 @@ private void OnControllerUpdate(WebXRControllerData controllerData)
...
+#else
             transform.localRotation = controllerData.rotation * controllerData.gripRotation;
             transform.localPosition = controllerData.rotation * (controllerData.position + controllerData.gripPosition);
+#endif
           }
           else
           {
+#if UNITY_2022_3_OR_NEWER
+            transform.SetLocalPositionAndRotation(controllerData.position, 
+                controllerData.rotation);
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRController.cs]:
```suggestion
              controllerData.rotation);
```

File: Packages/webxr/Runtime/Scripts/WebXRController.cs [hunk 16]
```diff
@@ -428,13 +438,23 @@ private void HandleOculusLinkBug(WebXRControllerData controllerData)
       gripPosition = controllerData.gripPosition;
       if (alwaysUseGrip)
       {
+#if UNITY_2022_3_OR_NEWER
+        transform.SetLocalPositionAndRotation(controllerData.rotation * (controllerData.position + controllerData.gripPosition), 
+            controllerData.rotation * controllerData.gripRotation);
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRController.cs]:
```suggestion
          controllerData.rotation * controllerData.gripRotation);
```

File: Packages/webxr/Runtime/Scripts/WebXRController.cs [hunk 17]
```diff
@@ -428,13 +438,23 @@ private void HandleOculusLinkBug(WebXRControllerData controllerData)
...
+#else
         transform.localRotation = controllerData.rotation * controllerData.gripRotation;
         transform.localPosition = controllerData.rotation * (controllerData.position + controllerData.gripPosition);
+#endif
       }
       else
       {
+#if UNITY_2022_3_OR_NEWER
+        transform.SetLocalPositionAndRotation(controllerData.position, 
+            controllerData.rotation * oculusOffsetRay);
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRController.cs]:
```suggestion
          controllerData.rotation * oculusOffsetRay);
```

File: Packages/webxr/Runtime/Scripts/WebXRController.cs [hunk 18]
```diff
@@ -468,10 +488,13 @@ private void OnHandUpdateInternal(WebXRHandData handData)
         }
         SetControllerActive(false);
         SetHandActive(true);
-
+#if UNITY_2022_3_OR_NEWER
+        transform.SetLocalPositionAndRotation(handData.joints[0].position, 
+            handData.joints[0].rotation);
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRController.cs]:
```suggestion
          handData.joints[0].rotation);
```

**Review by De-Panther — COMMENTED** [review]
Looks great, just some indentation changes needed

File: Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs [hunk 19]
```diff
@@ -183,19 +183,17 @@ public void SetUseCollidersForHandJoints(bool value)
...
         }
 #if WEBXR_INPUT_PROFILES
         if (handModelJoints.ContainsKey(i))
         {
-          var collider = handModelJoints[i].GetComponent<Collider>();
-          if (collider != null)
+          if (handModelJoints[i].TryGetComponent<Collider>(out var collider))
           {
-            collider.enabled = useCollidersForHandJoints;
+              collider.enabled = useCollidersForHandJoints;
```
Comment (De-Panther) [inline Packages/webxr-interactions/Runtime/Scripts/ControllerInteraction.cs]:
```suggestion
            collider.enabled = useCollidersForHandJoints;
```

File: Packages/webxr/Runtime/Scripts/WebXRController.cs [hunk 20]
```diff
@@ -371,13 +371,23 @@ private void OnControllerUpdate(WebXRControllerData controllerData)
...
           }
           else
           {
+#if UNITY_2022_3_OR_NEWER
+            transform.SetLocalPositionAndRotation(controllerData.position, 
+              controllerData.rotation);
+#else
             transform.localRotation = controllerData.rotation;
-            transform.localPosition = controllerData.position;
+            transform.localPosition = controllerData.position;
```
Comment (De-Panther) [inline Packages/webxr/Runtime/Scripts/WebXRController.cs]:
```suggestion
            transform.localPosition = controllerData.position;
```

**Review by De-Panther — COMMENTED** [review]
More indentation changes

**Review by De-Panther — APPROVED** [review]

**Comment by De-Panther** [comment]
🎉

**Comment by smitdylan2001** [comment]
Thanks for making the changes yourself!
Doing some awesome work with the exporter

**Comment by smitdylan2001** [comment]
@De-Panther 

It would be possible to use a custom # define at the top of each script with all the version checks, so 2021LTS users also can use the new feature without changing script. This might start to look even more cluttered however. If you think it's a good idea I could add it to a new PR.

Info:
https://forum.unity.com/threads/minor-version-or_newer-compilation.1445305

**Comment by De-Panther** [comment]
The issue is that it's both on 2021.3.11 and up in the 2021.3.x cycle, but then skips 2022.1.x and 2022.2.x cycles.
Maybe setting it in the `asmdef` file would look better?
https://docs.unity3d.com/Manual/ScriptCompilationAssemblyDefinitionFiles.html
"Defining symbols based on Unity and project package versions"
"Resource: choose Unity or the package or module that must be installed in order for this symbol to be defined"

**Comment by smitdylan2001** [comment]
Oh that's a big oversight of me indeed!
I am looking into the asmdef defines, but am unsure how to make it be so selectable.
For now maybe keeping it in 2022.3 is the best option unless a good solution pops up

**Comment by De-Panther** [comment]
For now 2022.3 is ok.
That's why I approved and merged :)
Thanks again

**Pull request merged**